distributed/scheduler.py

```diff
@@ -2229,6 +2223,12 @@ def transition_memory_released(self, key, stimulus_id, safe: bool = False):
             elif ts.who_wants or ts.waiters:
                 recommendations[key] = "waiting"
 
+            for dts in ts.waiters:
+                if dts.state in ("no-worker", "processing"):
+                    recommendations[dts.key] = "waiting"
+                elif dts.state == "waiting":
+                    dts.waiting_on.add(ts)
```

Comment:
> To confirm: the reason for moving this code later is because transitions are insertion-ordered,

Yes, dictionaries are insertion ordered and popitem pops from the end

```python
In [1]: adict = {}

In [2]: for x in range(10):
   ...:     adict[x] = x
   ...:

In [3]: adict.popitem()
Out[3]: (9, 9)
```

> I wonder if this block should even be in transition_memory_released at all?

Yes, it should. If a task that was in memory is no longer in memory and a waiter, i.e. a task to be executed, exists, we need to ensure that this waiter is released. These few lines allow task to be resilient to worker failures.

> Why shouldn't this part be done in transition_released_waiting

This transition is not there to reset/forget/release something but it schedules something for compute

> and transition_released_forgotten

This transition should only be triggered after the scheduler deletes the entire graph. This should only ever have any scheduling consequences if there are multiple graphs scheduled that share keys. Otherwise this is simply a sophisticated `pop task`

> If there's some dependency structure in the recommendations (transition A has to happen before transition B), I'd think transition A should be responsible for recommending transition B, not that they should be mixed together.

They are. The property we're looking for here is that if there is a chain of transitions for a certain Task, we should finish this transition chain first before moving to the next task. this change ensures this for this specific transition chain. It should be possible to implement this fairly easily directly in `Scheduler._transitions`. I'll have a look